### ISSUE TITLE
aws_elasticache_replication_group: Error: cannot upgrade state, missing MigrateState function

### ISSUE THREAD

**Issue by akatrevorjay** [terraform-providers/terraform-provider-aws#8828]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
2019/05/31 22:24:54 [INFO] Terraform version: 0.12.0  
2019/05/31 22:24:54 [INFO] Go runtime version: go1.12.4
```

### Affected Resource(s)

* aws_elasticache_replication_group

### Actual Behavior

Upon `terraform plan`:

```
2019-05-31T22:20:01.933Z [INFO]  plugin: configuring client automatic mTLS
2019/05/31 22:20:01 [TRACE] vertex "var.cpu_statistic": starting visit (*terraform.NodeRootVariable)
2019-05-31T22:20:01.984Z [INFO]  plugin.terraform-provider-aws_v2.12.0_x4: configuring server automatic mTLS: timestamp=2019-05-31T22:20:01.984Z
aws_elasticache_subnet_group.blah-elasticache: Refreshing state... [id=blah]
aws_sns_topic.blah-elasticache: Refreshing state... [id=blah]
aws_security_group.blah-elasticache: Refreshing state... [id=blah]
aws_security_group_rule.blah-elasticache: Refreshing state... [id=blah]
aws_security_group_rule.blah-elasticache-outbound: Refreshing state... [id=blah]
aws_security_group_rule.blah-elasticache-vpn: Refreshing state... [id=blah]
2019/05/31 22:20:04 [ERROR] <root>: eval: *terraform.EvalReadState, err: cannot upgrade state, missing MigrateState function
2019/05/31 22:20:04 [ERROR] <root>: eval: *terraform.EvalSequence, err: cannot upgrade state, missing MigrateState function
2019/05/31 22:20:04 [TRACE] [walkRefresh] Exiting eval tree: aws_elasticache_replication_group.blah

Error: cannot upgrade state, missing MigrateState function

```

### Steps to Reproduce

For me, `terraform plan`, but you'll likely need a similar state file I assume.

That said I don't expect this to be a non-obvious solution.

The state file in question is from `0.11.7`, so by running this it is updating it in memory.

**Comment by akatrevorjay** [comment]
I'm currently looking into this, but I wanted to make the issue first, as it may be an obvious easy fix for you. Thanks!

**Comment by bflad** [comment]
Hi @akatrevorjay 👋 Yikes that is unexpected behavior. Thanks for reporting it. Can you confirm if its also an issue with Terraform 0.12.1?

**Comment by bflad** [comment]
As a point of reference, I was not able to reproduce this by doing the following:

```hcl
provider "aws" {
  region  = "us-east-2"
  version = "2.13.0"
}

resource "aws_elasticache_replication_group" "example" {
  replication_group_id          = "issue-8828"
  replication_group_description = "https://github.com/terraform-providers/terraform-provider-aws/issues/8828"
  node_type                     = "cache.t2.micro"
  number_cache_clusters         = 1
  port                          = 6379
}
```

```console
$ terraform0.11.14 apply
...
Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
$ terraform version
Terraform v0.12.1
+ provider.aws v2.13.0
$ terraform apply
aws_elasticache_replication_group.example: Refreshing state... [id=issue-8828]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

The state file was showing schema version 1 from the original apply in Terraform 0.11, which probably explains why Terraform 0.12 didn't show an error in my quick testing here. My hunch is that I'll need to create the resource with version 1.13.0 (one! dot thirteen), which was the release before the inclusion of `SchemaVersion: 1` without `MigrateState` in this commit: https://github.com/terraform-providers/terraform-provider-aws/commit/681377a76daf43849b5e0ae8aa8ca3cc5b7da430#diff-542fde68a94ae5a5dea9eb1f63270320R121

@akatrevorjay can you confirm if this Terraform state was ever applied with a version of the Terraform AWS Provider post 1.14.0 (one! dot fourteen) prior to attempting an upgrade to Terraform 0.12? I'm going to try going back to then myself right now, but the confirmation either which way would be helpful. Thanks.

**Comment by bflad** [comment]
I've locally confirmed that the following sequence has the error:

* `terraform apply` with Terraform Core 0.11.14 and Terraform AWS Provider 1.13.0 (before `SchemaVersion: 1`) does not show a schema version in the Terraform state, which means the default schema version of `0`
* `terraform init && terraform plan` with Terraform Core 0.12.1 and Terraform AWS Provider 2.13.0 shows error:

```
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.


Error: cannot upgrade state, missing MigrateState function
```

And that the following sequence does not error:

* `terraform apply` with Terraform Core 0.11.14 and Terraform AWS Provider 1.13.0 (before `SchemaVersion: 1`) does not show a schema version in the Terraform state, which means the default schema version of `0`
* `terraform init && terraform apply` (no operation update) with Terraform Core 0.11.14 and Terraform AWS Provider 1.14.0 shows `"schema_version": "1"` in the Terraform state
* `terraform init && terraform plan` with Terraform Core 0.12.1 and Terraform AWS Provider 2.13.0 shows no errors

So while we should fix the code to include the missing state migration so a workaround isn't required, it does seem like operators who have updated their Terraform AWS Provider and applied in the last year should not hit this, so that's good. 😅 I can also confirm a successful workaround with applying on the latest version of the Terraform AWS Provider (2.13.0) as well before attempting any Terraform 0.12 operations so any successful `terraform refresh` or `terraform apply` with 1.14.0+ on Terraform 0.11 should do the trick to make Terraform 0.12 happy. 👍

**Comment by bflad** [comment]
The fix for this within the resource has been merged and will release with version 2.15.0 of the Terraform AWS Provider, likely later this week. That said, the upstream `missing MigrateState function` error is likely to be removed in the next version (v0.12.2), to match the previous behavior where `MigrateState` was not required and would allow the schema version update.
